### PR TITLE
fix(cache)!: handle failed requests to dsp-api (DSP-402)

### PR DESCRIPTION
**Pull request by tobiasschweizer**
DSP-402

This PR refactors the logic implemented in `GenericCache` so it handles errors correctly. Before, errors were cached and stayed in the cache. Now, failed requests are retried on each new subscription.

Previous to this PR, `GenericCache` created an `AsyncSubject` for each requested item that it wrote to once it retrieved the item  from DSP-API. It subscribed to the source Observable to do that and once the information was there, the `AsyncSubject` was completed. The effect of this was turning a unicast Observable  (the request to DSP-API) into a multicast Observable (the `AsyncSubject`). Although this worked the approach failed when dealing with errors since a failed Observable was cached and on each subscription the error was simply returned (forever).

This PR uses the `shareReplay` operator to turn the source Observable into a Subject. This approach is called [multicasting](https://netbasal.com/understanding-rxjs-multicast-operators-77b3f60af0a2).

The PR has the breaking flag because the cache's type changed from `AsyncSubject` to simply `Observable`. This affects all code that explicitly uses `AsyncSubject`. This code has to be refactored to use `Observable` instead:

- dsp-ui: `UserService`, `PropertyToolbarComponent`, `ViewerPlaygroundComponent`:
   user.service.spec.ts:
   ```ts
          const userCacheSpy = spyOn(service['_userCache'], 'getUser').and.callFake(
            () => {
                const user = MockUsers.mockUser();

                return of(user.body);
            }
        );
   ```
   and: property-toolbar.component.spec.ts, viewer-playground.component.spec.ts 

- beol: `ReadListValueComponent`   
   read-list-value.component.spec.ts:
   ```ts
        (listCacheServiceSpy.v2.listNodeCache as jasmine.SpyObj<ListNodeV2Cache>).getNode.and.callFake((nodeIri) => {
            const node = new ListNodeV2();
            node.id = 'nodeIri';
            node.label = 'test' + nodeIri;

            return of(node);
        });
   ```

### CODE REVIEW

**Comment by tobiasschweizer** [comment]
Good reads for myself:

- https://rxjs-dev.firebaseapp.com/api/operators/catchError
- https://blog.angular-university.io/rxjs-error-handling/
- https://www.learnrxjs.io/learn-rxjs/operators/error_handling/catch

**Comment by tobiasschweizer** [comment]
I figured out that `throwError` returns an `Observable` and [immediately emits an error notification](https://rxjs-dev.firebaseapp.com/api/index/function/throwError).

**Review by mdelez — APPROVED** [review]
Works well :) Thanks!

**Review by Vijeinath — APPROVED** [review]
Looks good for me

File: src/cache/GenericCache.ts [hunk 1]
```diff
@@ -14,103 +15,78 @@ export abstract class GenericCache<T> {
...
+        // also if the request has not finished yet
+        this.cache[key] = this.requestItemFromKnora(key, isDependency).pipe(
+            // only one item is expected
+            take(1),
+            // DSP-API may return several elements for the request (optimization)
+            tap((items: T[]) => {
+                if (items.length === 0) throw Error("No items returned from DSP-API for " + key);
+
+                // the first item is expected to be the requested item
+                if (key !== this.getKeyOfItem(items[0])) throw Error("First item of items returned from DSP-API is expected to be " + key);
```
[the remark this answers is not in the document]
Comment (tobiasschweizer) [reply src/cache/GenericCache.ts:47]: Good hint, will do that!

[the remark this answers is not in the document]
Comment (tobiasschweizer) [reply src/cache/GenericCache.ts]: fixed in https://github.com/dasch-swiss/dsp-js-lib/pull/210/commits/9cea53e56e441d68ceff929120722095294784cb

**Pull request merged**